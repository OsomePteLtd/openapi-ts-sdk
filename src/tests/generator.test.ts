import { join } from 'path';
import { generate, generateFromOpenApiSpecs } from './generate';

it('Basic Scenario', async () => {
  const endpoints = `
    @token GET /companies?CompanyQuery
      => CompaniesResponse

    @token POST /companies CompanyNewRequest
      => CompanyResponse

    @token GET /companies/:id
      => CompanyResponse

    @token PATCH /companies/:id CompanyUpdateRequest
      => CompanyResponse

    @token DELETE /companies/:id
      => SuccessResponse

    @token POST /companies/:id/archive
      => SuccessResponse
  `;
  const models = `
    SuccessResponse { success: b }
    Company { id: i, name: s, size: i, meta?: o }
    CompanyNew { name: s, size: i }
    CompanyUpdate { name?: s, size?: i }
    CompanyQuery { page?: i, perPage?: i }
    CompaniesResponse { companies: Company[], totalCount: i }
    CompanyResponse { company: Company }
    CompanyNewRequest { company: CompanyNew }
    CompanyUpdateRequest { company: CompanyUpdate }
  `;
  const { clientSource, typesSource, schemasSource } = await generate({
    endpoints,
    models,
  });
  expect(clientSource).toMatchSnapshot();
  expect(typesSource).toMatchSnapshot();
  expect(schemasSource).toMatchSnapshot();
});

it('Anonymous Types', async () => {
  const endpoints = `
    @token GET /billing/invoices/:id:i
      => { invoice: { id: i } }

    @token GET /billing/invoices/
      => { invoices: { id: i }[] }

    @token POST /companies { company: CompanyNew }
      => { company: Company }
  `;
  const models = `
    Company { id: i, name: s, size: i }
    CompanyNew { name: s, size: i }
  `;
  const { clientSource, typesSource } = await generate({ endpoints, models });
  expect(clientSource).toMatchSnapshot();
  expect(typesSource).toMatchSnapshot();
});

it('Enums', async () => {
  const endpoints = `
    @token GET /test
      => s
  `;
  const models = `
    CompanyDealStatus (
      new |
      won |
      lost |
      lostClient |
      test |
      junk
    )
    CompanyType (
      Branch Office Singapore |
      Exempt Private Company Limited by Shares (Pte. Ltd.) |
      Private Limited Company, use of 'Limited' exemption
    )
    CompanyTypeWithSimilarValues (
      Public limited company |
      Public Limited Company
    )
    ProcessDefinitionKey (
      cs-incorporation |
      cs-incorporation-pte-ltd-local |
      cs-appointment-of-new-secretary
    )
    BankAccountStatus (
      ACTIVE |
      ARCHIVED |
    )
  `;
  const { typesSource } = await generate({ endpoints, models });
  expect(typesSource).toMatchSnapshot();
});

it('Dashes and Underscores', async () => {
  const endpoints = `
    GET /user-roles/:role_id
      => { user_role: UserRole }
  `;
  const models = `
    UserRole { id: i, name: s }
  `;
  const { clientSource, typesSource } = await generate({ endpoints, models });
  expect(clientSource).toMatchSnapshot();
  expect(typesSource).toMatchSnapshot();
});

it('Schema Prefix', async () => {
  const endpoints = `
    GET /users/:id
      => { user: User }
  `;
  const models = `
    User { id: i, name: s }
  `;
  const { schemasSource } = await generate({
    endpoints,
    models,
    prefix: 'myFancyPrefix',
  });
  expect(schemasSource).toMatchSnapshot();
});

it('Id Path Parameter', async () => {
  const endpoints = `
    @token GET /companies/:companyId
      => CompanyResponse
  `;
  const models = `
    Company { id: i, name: s, size: i, meta?: o }
    CompanyResponse { company: Company }
  `;
  const { clientSource } = await generate({
    endpoints,
    models,
  });
  expect(clientSource).toMatchSnapshot();
});

it('Typed Schemas', async () => {
  const endpoints = `
    GET /users/:id
      => { user: User }
  `;
  const models = `
    User { id: i, name: s }
  `;
  const { schemasSource } = await generate({
    endpoints,
    models,
    typedSchemas: true,
  });
  expect(schemasSource).toMatchSnapshot();
});

it('Fakes', async () => {
  const models = `
    WithoutId { name: s }
    User { id: i, email: s }
    Company { id: i, name: s }
  `;
  const { fakesSource } = await generate({ endpoints: '', models });
  expect(fakesSource).toMatchSnapshot();
});

describe('OpenAPI V3', () => {
  it('Basic Scenario', async () => {
    const { clientSource, typesSource, schemasSource } =
      await generateFromOpenApiSpecs({
        files: [join(__dirname, 'assets', 'openapi-v3', 'petstore.json')],
      });
    expect(clientSource).toMatchSnapshot();
    expect(typesSource).toMatchSnapshot();
    expect(schemasSource).toMatchSnapshot();
  });
});

describe('OpenAPI V3_1', () => {
  it('Basic Scenario', async () => {
    const { clientSource, typesSource, schemasSource } =
      await generateFromOpenApiSpecs({
        files: [join(__dirname, 'assets', 'openapi-v3_1', 'schema.json')],
      });
    expect(clientSource).toMatchSnapshot();
    expect(typesSource).toMatchSnapshot();
    expect(schemasSource).toMatchSnapshot();
  });
});
