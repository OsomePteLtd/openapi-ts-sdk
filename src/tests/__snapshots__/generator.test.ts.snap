// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Anonymous Types 1`] = `
"import { AxiosRequestConfig, AxiosError } from 'axios';

import { GetRequestOptions, SdkOptions } from './options';
import { SdkRequester } from './requester';
import * as types from './types';

export function createSdkClient(options: SdkOptions) {
  const requester = new SdkRequester(options);
  return {
    setAuthToken(authToken: string | undefined) {
      requester.setAuthToken(authToken);
    },
    setErrorHandler(handler: (error: AxiosError) => void) {
      requester.setErrorHandler(handler);
    },
    get(path: string, query?: object, getOptions?: GetRequestOptions) {
      return requester.get(path, query, getOptions);
    },
    post(path: string, data?: object) {
      return requester.post(path, data);
    },
    put(path: string, data?: object) {
      return requester.put(path, data);
    },
    patch(path: string, data?: object) {
      return requester.patch(path, data);
    },
    delete(path: string, data?: object) {
      return requester.delete(path, data);
    },
    request(options: AxiosRequestConfig) {
      return requester.request(options);
    },
    billing: {
      invoices: {
        id: (id: number | string) => ({
          get(
            query?: any,
            options?: GetRequestOptions,
          ): Promise<types.AgGetBillingInvoicesByIdResponse> {
            return requester.get(\`/billing/invoices/\${id}\`, query, options);
          },
        }),
      },
    },
    companies: {
      post(
        data: types.AgPostCompaniesRequest,
      ): Promise<types.AgPostCompaniesResponse> {
        return requester.post(\`/companies\`, data);
      },
    },
  };
}

export type SdkClient = ReturnType<typeof createSdkClient>;
"
`;

exports[`Anonymous Types 2`] = `
"export interface AgGetBillingInvoicesByIdResponse {
  invoice: {
    id: number;
  };
}
export interface AgGetBillingInvoicesResponse {
  invoices: {
    id: number;
  }[];
}
export interface AgPostCompaniesRequest {
  company: CompanyNew;
}
export interface AgPostCompaniesResponse {
  company: Company;
}
export interface Company {
  id: number;
  name: string;
  size: number;
}
export interface CompanyNew {
  name: string;
  size: number;
}
"
`;

exports[`Basic Scenario 1`] = `
"import { AxiosRequestConfig, AxiosError } from 'axios';

import { GetRequestOptions, SdkOptions } from './options';
import { SdkRequester } from './requester';
import * as types from './types';

export function createSdkClient(options: SdkOptions) {
  const requester = new SdkRequester(options);
  return {
    setAuthToken(authToken: string | undefined) {
      requester.setAuthToken(authToken);
    },
    setErrorHandler(handler: (error: AxiosError) => void) {
      requester.setErrorHandler(handler);
    },
    get(path: string, query?: object, getOptions?: GetRequestOptions) {
      return requester.get(path, query, getOptions);
    },
    post(path: string, data?: object) {
      return requester.post(path, data);
    },
    put(path: string, data?: object) {
      return requester.put(path, data);
    },
    patch(path: string, data?: object) {
      return requester.patch(path, data);
    },
    delete(path: string, data?: object) {
      return requester.delete(path, data);
    },
    request(options: AxiosRequestConfig) {
      return requester.request(options);
    },
    companies: {
      id: (id: number | string) => ({
        archive: {
          post(data: any = {}): Promise<types.SuccessResponse> {
            return requester.post(\`/companies/\${id}/archive\`, data);
          },
        },
        get(
          query?: any,
          options?: GetRequestOptions,
        ): Promise<types.CompanyResponse> {
          return requester.get(\`/companies/\${id}\`, query, options);
        },
        patch(
          data: types.CompanyUpdateRequest,
        ): Promise<types.CompanyResponse> {
          return requester.patch(\`/companies/\${id}\`, data);
        },
        delete(data: any = {}): Promise<types.SuccessResponse> {
          return requester.delete(\`/companies/\${id}\`, data);
        },
      }),
      get(
        query?: types.CompanyQuery,
        options?: GetRequestOptions,
      ): Promise<types.CompaniesResponse> {
        return requester.get(\`/companies\`, query, options);
      },
      post(data: types.CompanyNewRequest): Promise<types.CompanyResponse> {
        return requester.post(\`/companies\`, data);
      },
    },
  };
}

export type SdkClient = ReturnType<typeof createSdkClient>;
"
`;

exports[`Basic Scenario 2`] = `
"export interface CompaniesResponse {
  companies: Company[];
  totalCount: number;
}
export interface Company {
  id: number;
  name: string;
  size: number;
}
export interface CompanyNew {
  name: string;
  size: number;
}
export interface CompanyNewRequest {
  company: CompanyNew;
}
export interface CompanyQuery {
  page?: number;
  perPage?: number;
}
export interface CompanyResponse {
  company: Company;
}
export interface CompanyUpdate {
  name?: string;
  size?: number;
}
export interface CompanyUpdateRequest {
  company: CompanyUpdate;
}
export interface SuccessResponse {
  success: boolean;
}
"
`;

exports[`Basic Scenario 3`] = `
"export const schemas = {
  SuccessResponse: {} as any,
  Company: {} as any,
  CompanyNew: {} as any,
  CompanyUpdate: {} as any,
  CompanyQuery: {} as any,
  CompaniesResponse: {} as any,
  CompanyResponse: {} as any,
  CompanyNewRequest: {} as any,
  CompanyUpdateRequest: {} as any,
};

schemas.SuccessResponse = Object.assign(schemas.SuccessResponse, {
  type: 'object',
  properties: { success: { type: 'boolean' } },
  required: ['success'],
});

schemas.Company = Object.assign(schemas.Company, {
  type: 'object',
  properties: {
    id: { type: 'integer' },
    name: { type: 'string' },
    size: { type: 'integer' },
  },
  required: ['id', 'name', 'size'],
});

schemas.CompanyNew = Object.assign(schemas.CompanyNew, {
  type: 'object',
  properties: { name: { type: 'string' }, size: { type: 'integer' } },
  required: ['name', 'size'],
});

schemas.CompanyUpdate = Object.assign(schemas.CompanyUpdate, {
  type: 'object',
  properties: { name: { type: 'string' }, size: { type: 'integer' } },
});

schemas.CompanyQuery = Object.assign(schemas.CompanyQuery, {
  type: 'object',
  properties: { page: { type: 'integer' }, perPage: { type: 'integer' } },
});

schemas.CompaniesResponse = Object.assign(schemas.CompaniesResponse, {
  type: 'object',
  properties: {
    companies: { type: 'array', items: schemas.Company },
    totalCount: { type: 'integer' },
  },
  required: ['companies', 'totalCount'],
});

schemas.CompanyResponse = Object.assign(schemas.CompanyResponse, {
  type: 'object',
  properties: { company: schemas.Company },
  required: ['company'],
});

schemas.CompanyNewRequest = Object.assign(schemas.CompanyNewRequest, {
  type: 'object',
  properties: { company: schemas.CompanyNew },
  required: ['company'],
});

schemas.CompanyUpdateRequest = Object.assign(schemas.CompanyUpdateRequest, {
  type: 'object',
  properties: { company: schemas.CompanyUpdate },
  required: ['company'],
});

Object.keys(schemas).forEach((name) => {
  schemas[name as keyof typeof schemas].$id = name;
});
"
`;

exports[`Dashes and Underscores 1`] = `
"import { AxiosRequestConfig, AxiosError } from 'axios';

import { GetRequestOptions, SdkOptions } from './options';
import { SdkRequester } from './requester';
import * as types from './types';

export function createSdkClient(options: SdkOptions) {
  const requester = new SdkRequester(options);
  return {
    setAuthToken(authToken: string | undefined) {
      requester.setAuthToken(authToken);
    },
    setErrorHandler(handler: (error: AxiosError) => void) {
      requester.setErrorHandler(handler);
    },
    get(path: string, query?: object, getOptions?: GetRequestOptions) {
      return requester.get(path, query, getOptions);
    },
    post(path: string, data?: object) {
      return requester.post(path, data);
    },
    put(path: string, data?: object) {
      return requester.put(path, data);
    },
    patch(path: string, data?: object) {
      return requester.patch(path, data);
    },
    delete(path: string, data?: object) {
      return requester.delete(path, data);
    },
    request(options: AxiosRequestConfig) {
      return requester.request(options);
    },
    user_roles: {
      role_id: (role_id: number | string) => ({
        get(
          query?: any,
          options?: GetRequestOptions,
        ): Promise<types.AgGetUserRolesByRoleIdResponse> {
          return requester.get(\`/user-roles/\${role_id}\`, query, options);
        },
      }),
    },
  };
}

export type SdkClient = ReturnType<typeof createSdkClient>;
"
`;

exports[`Dashes and Underscores 2`] = `
"export interface AgGetUserRolesByRoleIdResponse {
  user_role: UserRole;
}
export interface UserRole {
  id: number;
  name: string;
}
"
`;

exports[`Enums 1`] = `
"export enum BankAccountStatus {
  active = 'ACTIVE',
  archived = 'ARCHIVED',
  empty = '',
}
export enum CompanyDealStatus {
  new = 'new',
  won = 'won',
  lost = 'lost',
  lostClient = 'lostClient',
  test = 'test',
  junk = 'junk',
}
export enum CompanyType {
  branchOfficeSingapore = 'Branch Office Singapore',
  exemptPrivateCompanyLimitedBySharesPteLtd = 'Exempt Private Company Limited by Shares (Pte. Ltd.)',
  privateLimitedCompanyUseOfLimitedExemption = \\"Private Limited Company, use of 'Limited' exemption\\",
}
export type CompanyTypeWithSimilarValues =
  | 'Public limited company'
  | 'Public Limited Company';
export enum ProcessDefinitionKey {
  csIncorporation = 'cs-incorporation',
  csIncorporationPteLtdLocal = 'cs-incorporation-pte-ltd-local',
  csAppointmentOfNewSecretary = 'cs-appointment-of-new-secretary',
}
"
`;
